### PR TITLE
Application improvements

### PR DESCRIPTION
**Pull request by manicminer**
- Case insensitive matching on UUIDs for passwords and certificates
- Role/scope validation also checks IDs now that users must set it themselves
- Extensive whole-resource validation when `sign_in_audience` is changed to include personal accounts, to avoid breakage
- Updates Hamilton to v0.20.0 for fixes, features and improved odata query support

### New Applications Features
- Support additional properties in the `api` block
  - `accept_mapped_claims`
  - `known_client_applications`
  - `requested_access_token_version`
- Export the `disabled_by_microsoft` attribute
- Support the `logo_url`, `marketing_url`, `privacy_statement_url`, `support_url` and `terms_of_service_url` properties
- Support the `device_only_auth_enabled` property
- Support the `oauth2_post_response_required` property
- Export the `publisher_domain` attribute (this is readonly, awaiting feedback on how we can set this)
- Support the `public_block` block
- Support the `single_page_application` block
- Export two computed typemap attributes: `app_role_ids` and `oauth2_permission_scope_ids` to alleviate typeset handling with config boilerplate

Closes: #188
Closes: #244
Closes: #286
Closes: #320 
Closes: #436
Closes: #462
Closes: #469
Closes: #477

### CODE REVIEW

File: internal/services/applications/application_data_source.go
Comment (katbyte) [inline internal/services/applications/application_data_source.go:46]: i don't think we need status?
```suggestion
			"disabled_by_microsoft": {
```

File: internal/services/applications/application_data_source.go
Comment (katbyte) [inline internal/services/applications/application_data_source.go:66]: should we use this for consistency with other bools?
```suggestion
						"mapped_claims_enabled": {
```

File: internal/services/applications/application_resource.go
Comment (katbyte) [inline internal/services/applications/application_resource.go:277]: info seems superfluous - should we just move these all to the top level isntead?

File: internal/services/applications/application_resource.go
Comment (katbyte) [inline internal/services/applications/application_resource.go:317]: this says "post allowed" vs the property name "response_required"

File: internal/services/applications/application_resource.go
Comment (katbyte) [inline internal/services/applications/application_resource.go:351]: makes sense to pull this out to a top level property `public_client_redirect_uris`

File: internal/services/applications/application_resource.go
Comment (katbyte) [inline internal/services/applications/application_resource.go:420]: and here? or will there be other properties there

**Review by katbyte — APPROVED** [review]
aside from a few comments on schema this looks good!

File: internal/services/applications/application_resource.go
Comment (manicminer) [reply internal/services/applications/application_resource.go:277]: Agreed, that makes sense

File: internal/services/applications/application_resource.go
Comment (manicminer) [reply internal/services/applications/application_resource.go:317]: This is a tricky one since the docs are ambiguous on the behaviour. I'll try to dig up the actual meaning and make it clearer one way or the other.

File: internal/services/applications/application_resource.go
Comment (manicminer) [reply internal/services/applications/application_resource.go:420]: Yeah I'm sort of anticipating additional properties over time in both the `public_client` and `single_page_application` blocks, so was trying to preempt future deprecations here

File: internal/services/applications/application_resource.go
Comment (manicminer) [reply internal/services/applications/application_resource.go:284]: It seems like this description, which was pulled from [application manifest docs](https://docs.microsoft.com/en-gb/azure/active-directory/develop/reference-app-manifest?WT.mc_id=Portal-Microsoft_AAD_RegisteredApps#oauth2requiredpostresponse-attribute), is the only documentation available on this one. Will follow up separately and try to find out more.

**Comment by manicminer** [comment]
@katbyte I've added some additional validation to try and avoid applying bad application configurations; could you take another look before merging? Thanks! :)

**Comment by manicminer** [comment]
This is unfortunately blocked due to a plugin SDK issue around block handling

**Comment by manicminer** [comment]
Test results

Screenshot 2021-07-04 at 19 20 08

File: internal/services/applications/application_data_source.go
Comment (katbyte) [inline internal/services/applications/application_data_source.go:352]: if this is a list shouldn't it be plural?
```suggestion
			"single_page_applications": {
```

File: internal/services/applications/application_resource.go
Comment (katbyte) [reply internal/services/applications/application_resource.go:304]: any update on this one?

**Review by katbyte — APPROVED** [review]
aside from a couple comments LGTM 🍰

File: internal/services/applications/application_data_source.go
Comment (manicminer) [reply internal/services/applications/application_data_source.go:352]: This block is intended to contain the settings for the application when it's functioning as a single page app (the API container is just `spa` but I felt it needed expanding)

File: internal/services/applications/application_resource.go
Comment (manicminer) [reply internal/services/applications/application_resource.go:304]: Not as yet, have reached out to get some clarity on this field

Comment (manicminer) [reply internal/services/applications/application_resource.go:314]: I've raised this separately as #491 against the 2.0.0 milestone in order to unblock the PR stack :)

**Pull request merged**
